### PR TITLE
unit tests: Add mz_ore::test to initialize logging implicitly

### PR DESCRIPTION
**Pull request by def-**
It seems easy to forget to initialize logging. Unfortunately there is no easy way of doing this in Rust currently. See for example a recent Slack discussion: https://materializeinc.slack.com/archives/C0246GEHL8N/p1683754885066559

I'll rebase this once the reviews are through since it has huge conflict potential. 

### Motivation

   * This PR refactors existing code.

### Tips for reviewer

Review individual commits.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Review by benesch — COMMENTED** [review]
This is a fantastic idea! Let's have @guswynn or @petrosagg or some other Rust expert review the proc macro itself, but +100 on the idea from me.

Can we do a blanket find/replace on #[test] in the codebase, too? It seems that you've limited the swap for just those tests that `mz_ore::test::init_logging()` was called in, but really any test function that _doesn't_ call that is a recipe for trouble. Then we should also update STYLE.md to say that use of this macro is mandatory.

File: .github/CODEOWNERS [hunk 1]
```diff
@@ -84,6 +84,7 @@
 /src/stash-debug                    @MaterializeInc/surfaces
 /src/storage                        @MaterializeInc/storage
 /src/storage-client                 @MaterializeInc/storage
+/src/test-macro                     @MaterializeInc/qa
```
Comment (def-) [inline .github/CODEOWNERS:87]: @philip-stoev I hope this is ok, otherwise can also assign it to myself directly.

File: src/test-macro/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+use syn::ItemFn;
+use syn::Meta;
+use syn::NestedMeta;
+use syn::ReturnType;
+
+/// Based on <https://github.com/d-e-s-o/test-log>
+/// Copyright (C) 2019-2022 Daniel Mueller <deso@posteo.net>
+/// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+#[proc_macro_attribute]
```
Comment (def-) [inline src/test-macro/src/lib.rs:101]: This is the actually interesting part of this PR, the rest is just to actually use it.

File: src/test-macro/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+    let result = quote! {
+      #[#inner_test]
+      #(#attrs)*
+      #async_ fn #test_name() #ret {
+        #async_ fn test_impl() #ret {
+          #body
+        }
+
+        #logging_init
```
Comment (def-) [inline src/test-macro/src/lib.rs:163]: If we later have other things we want to run in each unit test, during setup or at the end, we can add them here in the future.

File: src/test-macro/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+pub fn test(attr: TokenStream, item: TokenStream) -> TokenStream {
+    let args = parse_macro_input!(attr as AttributeArgs);
+    let input = parse_macro_input!(item as ItemFn);
+
+    let inner_test = match args.as_slice() {
+        [] => parse_quote! { ::core::prelude::v1::test },
+        [NestedMeta::Meta(Meta::Path(path))] => quote! { #path },
+        [NestedMeta::Meta(Meta::List(list))] => quote! { #list },
+        _ => panic!("unsupported attributes supplied: {:?}", args),
+    };
```
[the remark this answers is not in the document]
Comment (def-) [reply src/test-macro/src/lib.rs:111]: It does, I used the `flavor = "multi_threads"` in some tests and haven't seen a problem. Example from the project I took this from: https://github.com/d-e-s-o/test-log/blob/f500323f947a4bfdcb015edbc814555ddb04a0fc/tests/mod.rs#L71

[the remark this answers is not in the document]
Comment (guswynn) [reply src/test-macro/src/lib.rs:111]: Oh then this looks good!

File: .github/CODEOWNERS (hunk 1 above)
Comment (philip-stoev) [reply .github/CODEOWNERS:87]: Yes, it is ok, thanks.

**Review by philip-stoev — APPROVED** [review]
Thank you. Can we get one of the linters to start rejecting incorrect direct use of the test macro? I believe @teskje was the last person that was trying to use a linter to prohibit certain constructs altogether.

**Review by nrainer-materialize — APPROVED** [review]

**Comment by def-** [comment]
I don't know how to create Rust-parser based warnings myself. I can easily do a regex though if that is sufficient.

**Comment by teskje** [comment]
Clippy's [`disallowed-macros`](https://rust-lang.github.io/rust-clippy/stable/index.html#disallowed_macros) config seems like it should work for this, but I tried for a bit and couldn't get it to flag `#[test]` or `#[tokio::test]`.

**Review by teskje — APPROVED** [review]
LGTM, though I'll also leave the proc macro review to more knowledgable folk.

I agree that it would be nice to have something in CI flagging use of non-mz test macros.

**Comment by def-** [comment]
`clippy` is unfortunately insufficient here, since it will complain about the location where the mz_ore::test macro is used, so we'd have to silence the warning everywhere, which defeats the purpose. I tried with:
```diff
diff --git a/clippy.toml b/clippy.toml
index 83b1777a60..c92db4b1ee 100644
--- a/clippy.toml
+++ b/clippy.toml
@@ -49,6 +49,9 @@ disallowed-methods = [
 disallowed-macros = [
     { path = "proptest::prop_oneof", reason = "use `proptest::strategy::Union::new` instead" },
     { path = "log::log", reason = "use the macros provided by `tracing` instead (#9992)" },
+    { path = "core::prelude::v1::test", reason = "use the `mz_ore::test` macro instead" },
+    # How to warn about tokio::test but not mz_ore::test(tokio::test)?
+    #{ path = "tokio::test", reason = "use the `mz_ore::test(tokio::test)` macro instead" },
 ]

 disallowed-types = [
diff --git a/src/test-macro/src/lib.rs b/src/test-macro/src/lib.rs
index 4748331929..8766be60d0 100644
--- a/src/test-macro/src/lib.rs
+++ b/src/test-macro/src/lib.rs
@@ -104,6 +104,7 @@ pub fn test(attr: TokenStream, item: TokenStream) -> TokenStream {
     let input = parse_macro_input!(item as ItemFn);

     let inner_test = match args.as_slice() {
+        #[allow(clippy::disallowed_macros)]
         [] => parse_quote! { ::core::prelude::v1::test },
         [NestedMeta::Meta(Meta::Path(path))] => quote! { #path },
         [NestedMeta::Meta(Meta::List(list))] => quote! { #list },
@@ -153,6 +154,7 @@ fn expand_wrapper(inner_test: &Tokens, wrappee: &ItemFn) -> TokenStream {
     let logging_init = expand_logging_init();

     let result = quote! {
+      // Allow access to core::macros::builtin::test inside this wrapper
       #[#inner_test]
       #(#attrs)*
       #async_ fn #test_name() #ret {
```

**Comment by philip-stoev** [comment]
Ok then so the new boilerplate code will desseminate using copy+paste then.

File: misc/lint/test-attribute.sh [hunk 5]
```diff
@@ -0,0 +1,50 @@
+#!/usr/bin/env bash
```
Comment (def-) [inline misc/lint/test-attribute.sh:1]: As requested, added a check. Sample output:
```
lint: error: test-attribute: src/ore/tests/future.rs: use of disallowed `#[tokio::test] attribute. Use the `#[mz_ore::test(tokio::test)]` attribute instead or add a `// allow(test-attribute)` comment
```

Comment (nrainer-materialize) [reply misc/lint/test-attribute.sh:1]: Cool! 🎉

**Comment by def-** [comment]
I'll set this to auto-merge unless someone complains and/or wants to review the macro.

**Pull request merged**